### PR TITLE
Quizmaster sees and corrects Team Answers

### PR DESCRIPTION
**Pull request by AmberWilkie**
PT Story: https://www.pivotaltracker.com/story/show/132561501

Changes proposed in this pull request:
- Quizmaster can mark answers right or wrong
- Quizmaster can "undo" an answer (that is, he can get the buttons back and resubmit - the "undo" button does not remove the `is_correct` value)
- Back button to go back to Questions page

What I have learned working on this feature:
- How to use `onClick` listeners in JQuery to find `this` and manipulate elements inside `this`
- How to use regex to find all of the elements with the same ID in the DOM

Screenshots:
screen shot 2016-10-22 at 11 42 18 am

Ready for review!

### CODE REVIEW

File: features/step_definitions/quizmaster_steps.rb [hunk 1]
```diff
@@ -2,3 +2,23 @@
...
+  question.answers.each do |response|
+    if response.body == answer
+      within("#answer_#{response.id}") { find("##{is_correct}").trigger('click') }
+    end
+  end
+end
+
+Then(/^"([^"]*)" should have "([^"]*)" correct (?:answer|answers)$/) do |team_name, count|
+  team = Team.find_by(name: team_name)
+  # binding.pry
```
Comment (tochman) [inline features/step_definitions/quizmaster_steps.rb]: pry

File: features/step_definitions/player_steps.rb [hunk 2]
```diff
@@ -47,6 +47,7 @@
   #expect(cookie.to_i).to eq team_id
 end
 
-And(/^there is a team named "([^"]*)"$/) do |name|
-  FactoryGirl.create(:team, name: name)
-end
\ No newline at end of file
+And(/^there is a team named "([^"]*)" playing "([^"]*)"$/) do |team_name, quiz_name|
+  quiz = Quiz.find_by(name: quiz_name)
+  Team.create(name: team_name, quiz: quiz)
```
Comment (tochman) [inline features/step_definitions/player_steps.rb]: Why not use `FactoryGirl`?

File: features/step_definitions/answer_steps.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Given(/^"([^"]*)" has answered question "([^"]*)" with "([^"]*)"$/) do |team_name, question, answer|
+  team = Team.find_by(name: team_name)
+  question = Question.find_by(body: question)
+  answer = Answer.create(team: team, question: question, body: answer)
```
Comment (tochman) [inline features/step_definitions/answer_steps.rb]: Why not use `FactoryGirl`?

File: features/quizmaster/view_quiz_page.feature [hunk 4]
```diff
@@ -5,7 +5,7 @@ Feature: As a Quizmaster
 
 Background:
   Given there is a quiz called "Trivia"
-  And there is a team named "Craft Academy"
+  And there is a team named "Craft Academy" playing "Trivia"
```
Comment (tochman) [inline features/quizmaster/view_quiz_page.feature]: Perhaps you should consider refactoring the step def for this to include setting the cookie (see below) and  minimizing exposure to what is going on in the background in the scenario? Perhaps it is not possible. Just a suggestion.

File: features/quizmaster/view_answers.feature [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  Given I press the "Send" button for question "1"
+  And "Amber Rocks!" has answered question "What is 2+2?" with "4"
+  And "Craft Academy" has answered question "What is 2+2?" with "Six"
+  When I click the "Correct" button
+  Then I should be on the correction page for "What is 2+2?"
+  When I click "right" on "4"
+  And I click "wrong" on "Six"
+  Then "Amber Rocks!" should have "1" correct answer
+  And "Craft Academy" should have "0" correct answers
+  And I should see "Undo?"
```
Comment (tochman) [inline features/quizmaster/view_answers.feature]: So adding the Undo functionality feels out of scope to be honest. Not a blocker but I wanted to point that out.

File: config/routes.rb [hunk 6]
```diff
@@ -4,6 +4,8 @@
     resources :quiz, controller: :quizzes, only: [:show]
     post '/quiz/:id', controller: :quizzes, action: :start_quiz
     post '/send_question', controller: :quizzes, action: :send_question
+    get 'quiz/answers/:id', controller: :quizzes, action: :correct_answers
+    get 'quiz/answers/mark/:id', controller: :quizzes, action: :mark_answers
```
Comment (tochman) [inline config/routes.rb]: No `/` ??

File: app/views/quizmaster/quizzes/answers.html.haml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+  .answers
+    - @question.answers.each do |answer|
+      %divr{id: "answer_#{answer.id}"}
+        = answer.body
+        %divr{id: "right_or_wrong_#{answer.id}", class: "answer_div"}
+          %divr{class: "correct_buttons"}
+            = link_to image_tag('x-box.png', class: 'square-button'), "/quizmaster/quiz/answers/mark/#{answer.id}?correct=false", id: "wrong", remote: true
+            = link_to image_tag('check-box.png', class: 'square-button'), "/quizmaster/quiz/answers/mark/#{answer.id}?correct=true", id: "right", remote: true
+
```
Comment (tochman) [inline app/views/quizmaster/quizzes/answers.html.haml]: As always I prefer if you use path helpers. You can still pass in params: `your_path(id: answer.id, correct: 'true')`. Have you tried that?

File: app/views/quizmaster/quizzes/answers.html.haml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  %h3= @question.answer
+  %hr/
+
+  .answers
+    - @question.answers.each do |answer|
+      %divr{id: "answer_#{answer.id}"}
+        = answer.body
+        %divr{id: "right_or_wrong_#{answer.id}", class: "answer_div"}
+          %divr{class: "correct_buttons"}
+            = link_to image_tag('x-box.png', class: 'square-button'), "/quizmaster/quiz/answers/mark/#{answer.id}?correct=false", id: "wrong", remote: true
```
Comment (tochman) [inline app/views/quizmaster/quizzes/answers.html.haml]: See comment below

File: app/controllers/quizmaster/quizzes_controller.rb [hunk 9]
```diff
@@ -18,11 +18,28 @@ def send_question
...
+
   def broadcast_content(content)
     # This method will broadcast content to Team
     BroadcastQuizJob.perform_now(content)
   end
 
+  def mark_answers
+    answer = Answer.find(params[:id])
+    if params[:correct] == 'true'
+      answer.is_correct = true
```
Comment (tochman) [inline app/controllers/quizmaster/quizzes_controller.rb]: Try `answer.update_attribute(:is_correct, true)` and save a line of code ;-)

File: app/controllers/quizmaster/quizzes_controller.rb [hunk 10]
```diff
@@ -18,11 +18,28 @@ def send_question
...
+      answer.save
+    elsif params[:correct] == 'false'
+      answer.is_correct = false
+      answer.save
```
Comment (tochman) [inline app/controllers/quizmaster/quizzes_controller.rb]: See comment above

File: app/views/quizmaster/quizzes/answers.html.haml (hunk 7 above)
Comment (AmberWilkie) [inline app/views/quizmaster/quizzes/answers.html.haml]: there's no path for this one...

File: features/quizmaster/view_answers.feature (hunk 5 above)
Comment (AmberWilkie) [inline features/quizmaster/view_answers.feature]: Yes it's out of scope.

File: app/views/quizmaster/quizzes/answers.html.haml (hunk 7 above)
Comment (tochman) [inline app/views/quizmaster/quizzes/answers.html.haml]: Why is that?

Comment (AmberWilkie) [inline app/views/quizmaster/quizzes/answers.html.haml]: er no path helper.
rails routes:
```
 Prefix Verb URI Pattern                                 Controller#Action
         quizmaster_quiz GET  /quizmaster/quiz/:id(.:format)              quizmaster/quizzes#show
              quizmaster POST /quizmaster/quiz/:id(.:format)              quizmaster/quizzes#start_quiz
quizmaster_send_question POST /quizmaster/send_question(.:format)         quizmaster/quizzes#send_question
                         GET  /quizmaster/quiz/answers/:id(.:format)      quizmaster/quizzes#correct_answers
                         GET  /quizmaster/quiz/answers/mark/:id(.:format) quizmaster/quizzes#mark_answers
        quiz_create_team POST /quiz/:quiz_id/create_team(.:format)        games#create_team
              quiz_index GET  /quiz(.:format)                             games#index
                    quiz GET  /quiz/:id(.:format)                         games#show
             access_quiz POST /access_quiz(.:format)                      games#access_quiz
```

File: app/controllers/quizmaster/quizzes_controller.rb (hunk 9 above)
Comment (AmberWilkie) [inline app/controllers/quizmaster/quizzes_controller.rb]: done

File: app/controllers/quizmaster/quizzes_controller.rb (hunk 10 above)
Comment (AmberWilkie) [inline app/controllers/quizmaster/quizzes_controller.rb]: done

File: config/routes.rb (hunk 6 above)
Comment (tochman) [inline config/routes.rb]: add `as: :correct_answers` or something to create a helper

File: app/views/quizmaster/quizzes/answers.html.haml (hunk 7 above)
Comment (tochman) [inline app/views/quizmaster/quizzes/answers.html.haml]: See my comment on `routes.rb`

File: config/routes.rb (hunk 6 above)
Comment (AmberWilkie) [inline config/routes.rb]: ah ha

File: features/quizmaster/view_quiz_page.feature (hunk 4 above)
Comment (AmberWilkie) [inline features/quizmaster/view_quiz_page.feature]: I did it like this to make sure the association between team and quiz were happening. Before this feature our Factory was assigning a random quiz for the team to have. It wasn't a problem before but was now (I think... it's been a while).

Comment (tochman) [inline features/quizmaster/view_quiz_page.feature]: Cool.

**Comment by tochman** [comment]
All good. Great job!

File: config/routes.rb [hunk 11]
```diff
@@ -4,6 +4,8 @@
     resources :quiz, controller: :quizzes, only: [:show]
     post '/quiz/:id', controller: :quizzes, action: :start_quiz
     post '/send_question', controller: :quizzes, action: :send_question
+    get 'quiz/answers/:id', controller: :quizzes, action: :correct_answers, as: :correct_answers
+    get 'quiz/answers/mark/:id', controller: :quizzes, action: :mark_answers, as: :mark_answers
   end
```
Comment (AmberWilkie) [inline config/routes.rb]: The route didn't seem to need the `/`...?

Comment (tochman) [inline config/routes.rb]: I'm a bit surprised by that. But okay.
